### DISCUSSION TITLE
Prevent automatic semicolon insertion when using babel@parser

### DISCUSSION THREAD

**Discussion by meetesh06** [babel/babel#16746] [Q&A] [answered]
I am using babel@parser 7.25.2 to generate ASTs for further processing. 

The file of concern ([S7.9_A1.js](https://github.com/tc39/test262/blob/main/test/language/asi/S7.9_A1.js) ):
```javascript
//CHECK#2
var result = false;
label2: for (var i = 0; i <= 1; i++) {
  for (var j = 0; j <= 1; j++) {
    if (j === 0) {
      continue    // <-- No semicolon here...
      label2;
    } else {
      result = true;
    }
  }
}

if (result !== true) {
  throw new Test262Error('#2: Check continue statement for automatic semicolon insertion');
}
```
Code generated by Babel: 
```javascript
//CHECK#2
var result = false;
label2: for (var i = 0; i <= 1; i++) {
  for (var j = 0; j <= 1; j++) {
    if (j === 0) {
      continue;  // <--- This is causing the test to fail
      label2;
    } else {
      result = true;
    }
  }
}
if (result !== true) {
  throw new Test262Error('#2: Check continue statement for automatic semicolon insertion');
}
```
Babel configuration I am using to generate this file

```javascript
let presets: Array<Array<string | {}>> = [
            ["@babel/preset-env", { targets: "last 2 Chrome versions", modules: false }],
]

if (that.extension === 'ts' || that.extension === 'tsx') presets.push(['@babel/preset-typescript'])
const options = {
            cwd: that.projectBasePath,
            filename: that.filename,
            sourceType,
            ast: true,
            presets,
            sourceMaps: true,
            plugins: [
              "@babel/plugin-syntax-jsx"
            ],
}
babel.transformAsync(sourceCode, options).then(result => {...})
```
I see that this test has passed on the https://test262.fyi/# site, so I am pretty sure there is something wrong with my configuration. Any help is appreciated :)

**Comment by nicolo-ribaudo** [comment, marked as answer]
The input code and output code are equivalent, as per spec there is an implicit semicolon after that `continue`. Otherwise the test would fail, because if that `continue` points to `label2` that `result` would never be set to true.

**Reply by meetesh06** [reply]
Thanks, its my generated [code](https://evanw.github.io/source-map-visualization/#MTQ2NQBsYWJlbDE6IGZvciAodmFyIGkgPSAwOzspIHsKICBsZXQganMzJDIgPSAwOwogIGxldCBqczMkMSA9IGkgPD0ganMzJDI7CiAgaWYgKGpzMyQxKSB7CiAgICBicmVhazsKICB9CiAgewogICAgZm9yICh2YXIgaiA9IDA7OykgewogICAgICBsZXQganMzJDQgPSAxOwogICAgICBsZXQganMzJDMgPSBqIDw9IGpzMyQ0OwogICAgICBpZiAoanMzJDMpIHsKICAgICAgICBicmVhazsKICAgICAgfQogICAgICB7CiAgICAgICAgbGV0IGlmVGVzdCQ2ID0gMDsKICAgICAgICBsZXQgaWZUZXN0JDUgPSBqID09PSBpZlRlc3QkNjsKICAgICAgICBpZiAoaWZUZXN0JDUpIHsKICAgICAgICAgIGNvbnRpbnVlIGxhYmVsMTsKICAgICAgICB9IGVsc2UgewogICAgICAgICAgbGV0IGlmRmFsc2UkOCA9ICcjMTogQ2hlY2sgY29udGludWUgc3RhdGVtZW50IGZvciBhdXRvbWF0aWMgc2VtaWNvbG9uIGluc2VydGlvbic7CiAgICAgICAgICBsZXQgaWZGYWxzZSQ3ID0gbmV3IFRlc3QyNjJFcnJvcihpZkZhbHNlJDgpOwogICAgICAgICAgdGhyb3cgaWZGYWxzZSQ3OwogICAgICAgIH0KICAgICAgfQogICAgICBsZXQganMzJDkgPSBqKys7CiAgICB9CiAgfQogIGxldCBqczMkMTAgPSBpKys7Cn0KbGV0IHJlc3VsdCQxMSA9IGZhbHNlOwp2YXIgcmVzdWx0ID0gcmVzdWx0JDExOwpsYWJlbDI6IGZvciAodmFyIGkgPSAwOzspIHsKICBsZXQganMzJDEzID0gMTsKICBsZXQganMzJDEyID0gaSA8PSBqczMkMTM7CiAgaWYgKGpzMyQxMikgewogICAgYnJlYWs7CiAgfQogIHsKICAgIGZvciAodmFyIGogPSAwOzspIHsKICAgICAgbGV0IGpzMyQxNSA9IDE7CiAgICAgIGxldCBqczMkMTQgPSBqIDw9IGpzMyQxNTsKICAgICAgaWYgKGpzMyQxNCkgewogICAgICAgIGJyZWFrOwogICAgICB9CiAgICAgIHsKICAgICAgICBsZXQgaWZUZXN0JDE3ID0gMDsKICAgICAgICBsZXQgaWZUZXN0JDE2ID0gaiA9PT0gaWZUZXN0JDE3OwogICAgICAgIGlmIChpZlRlc3QkMTYpIHsKICAgICAgICAgIGNvbnRpbnVlOwogICAgICAgICAgbGFiZWwyOwogICAgICAgIH0gZWxzZSB7CiAgICAgICAgICBsZXQgaWZGYWxzZSQxOCA9IHRydWU7CiAgICAgICAgICByZXN1bHQgPSBpZkZhbHNlJDE4CiAgICAgICAgICBsZXQgaWZGYWxzZSQxOSA9IHJlc3VsdDsKICAgICAgICAgIGlmRmFsc2UkMTk7CiAgICAgICAgfQogICAgICB9CiAgICAgIGxldCBqczMkMjAgPSBqKys7CiAgICB9CiAgfQogIGxldCBqczMkMjEgPSBpKys7Cn0KbGV0IGlmVGVzdCQyMyA9IHRydWU7CmxldCBpZlRlc3QkMjIgPSByZXN1bHQgIT09IGlmVGVzdCQyMzsKaWYgKGlmVGVzdCQyMikgewogIGxldCBpZlRydWUkMjUgPSAnIzI6IENoZWNrIGNvbnRpbnVlIHN0YXRlbWVudCBmb3IgYXV0b21hdGljIHNlbWljb2xvbiBpbnNlcnRpb24nOwogIGxldCBpZlRydWUkMjQgPSBuZXcgVGVzdDI2MkVycm9yKGlmVHJ1ZSQyNSk7CiAgdGhyb3cgaWZUcnVlJDI0Owp9MjcyOQB7InZlcnNpb24iOjMsIm5hbWVzIjpbImxhYmVsMSIsImkiLCJqczMkMiIsImpzMyQxIiwiaiIsImpzMyQ0IiwianMzJDMiLCJpZlRlc3QkNiIsImlmVGVzdCQ1IiwiaWZGYWxzZSQ4IiwiaWZGYWxzZSQ3IiwiVGVzdDI2MkVycm9yIiwianMzJDkiLCJqczMkMTAiLCJyZXN1bHQkMTEiLCJyZXN1bHQiLCJsYWJlbDIiLCJqczMkMTMiLCJqczMkMTIiLCJqczMkMTUiLCJqczMkMTQiLCJpZlRlc3QkMTciLCJpZlRlc3QkMTYiLCJpZkZhbHNlJDE4IiwiaWZGYWxzZSQxOSIsImpzMyQyMCIsImpzMyQyMSIsImlmVGVzdCQyMyIsImlmVGVzdCQyMiIsImlmVHJ1ZSQyNSIsImlmVHJ1ZSQyNCJdLCJzb3VyY2VzIjpbIlM3LjlfQTEuanMiXSwic291cmNlc0NvbnRlbnQiOlsiLy8gQ29weXJpZ2h0IDIwMDkgdGhlIFNwdXRuaWsgYXV0aG9ycy4gIEFsbCByaWdodHMgcmVzZXJ2ZWQuXG4vLyBUaGlzIGNvZGUgaXMgZ292ZXJuZWQgYnkgdGhlIEJTRCBsaWNlbnNlIGZvdW5kIGluIHRoZSBMSUNFTlNFIGZpbGUuXG5cbi8qLS0tXG5pbmZvOiBDaGVjayBDb250aW51ZSBTdGF0ZW1lbnQgZm9yIGF1dG9tYXRpYyBzZW1pY29sb24gaW5zZXJ0aW9uXG5lczVpZDogNy45X0ExXG5kZXNjcmlwdGlvbjogVHJ5IHVzZSBjb250aW51ZSBcXG4gTGFiZWwgY29uc3RydWN0aW9uXG4tLS0qL1xuXG4vL0NIRUNLIzFcbmxhYmVsMTogZm9yICh2YXIgaSA9IDA7IGkgPD0gMDsgaSsrKSB7XG4gIGZvciAodmFyIGogPSAwOyBqIDw9IDE7IGorKykge1xuICAgIGlmIChqID09PSAwKSB7XG4gICAgICBjb250aW51ZSBsYWJlbDE7XG4gICAgfSBlbHNlIHtcbiAgICAgIHRocm93IG5ldyBUZXN0MjYyRXJyb3IoJyMxOiBDaGVjayBjb250aW51ZSBzdGF0ZW1lbnQgZm9yIGF1dG9tYXRpYyBzZW1pY29sb24gaW5zZXJ0aW9uJyk7XG4gICAgfVxuICB9XG59XG5cbi8vQ0hFQ0sjMlxudmFyIHJlc3VsdCA9IGZhbHNlO1xubGFiZWwyOiBmb3IgKHZhciBpID0gMDsgaSA8PSAxOyBpKyspIHtcbiAgZm9yICh2YXIgaiA9IDA7IGogPD0gMTsgaisrKSB7XG4gICAgaWYgKGogPT09IDApIHtcbiAgICAgIGNvbnRpbnVlXG4gICAgICBsYWJlbDI7XG4gICAgfSBlbHNlIHtcbiAgICAgIHJlc3VsdCA9IHRydWU7XG4gICAgfVxuICB9XG59XG5cbmlmIChyZXN1bHQgIT09IHRydWUpIHtcbiAgdGhyb3cgbmV3IFRlc3QyNjJFcnJvcignIzI6IENoZWNrIGNvbnRpbnVlIHN0YXRlbWVudCBmb3IgYXV0b21hdGljIHNlbWljb2xvbiBpbnNlcnRpb24nKTtcbn1cbiJdLCJtYXBwaW5ncyI6IkFBVUFBLE1BQU0sRUFBRSxLQUFLLElBQUlDLENBQUMsR0FBRyxDQUFDLElBQWU7RUFBUixJQUFBQyxLQUFDLEdBQUQsQ0FBQztFQUFOLElBQUFDLEtBQU0sR0FBTkYsQ0FBQyxJQUFJQyxLQUFDO0VBQU4sSUFBQUMsS0FBTSxFQUF0QjtJQUFnQjtFQVF4QjtFQVJxQztJQUNuQyxLQUFLLElBQUlDLENBQUMsR0FBRyxDQUFDLElBQWU7TUFBUixJQUFBQyxLQUFDLEdBQUQsQ0FBQztNQUFOLElBQUFDLEtBQU0sR0FBTkYsQ0FBQyxJQUFJQyxLQUFDO01BQU4sSUFBQUMsS0FBTSxFQUF0QjtRQUFnQjtNQU1oQjtNQU42QjtRQUNqQixJQUFBQyxRQUFDLEdBQUQsQ0FBQztRQUFQLElBQUFDLFFBQU8sR0FBUEosQ0FBQyxLQUFLRyxRQUFDO1FBQVgsSUFBSUMsUUFBTyxFQUFFO1VBQ1gsU0FBU1IsTUFBTTtRQUNqQixDQUFDLE1BQU07VUFDa0IsSUFBQVMsU0FBZ0UsR0FBaEUsZ0VBQWdFO1VBQWpGLElBQUFDLFNBQWtGLEdBQWxGLElBQUlDLFlBQVksQ0FBQ0YsU0FBZ0UsQ0FBQztVQUF4RixNQUFNQyxTQUFrRjtRQUMxRjtNQUNGO01BTndCLElBQUFFLEtBQUcsR0FBSFIsQ0FBQyxFQUFFO0lBTTNCO0VBQ0Y7RUFSZ0MsSUFBQVMsTUFBRyxHQUFIWixDQUFDLEVBQUU7QUFRbkM7QUFHYSxJQUFBYSxTQUFLLEdBQUwsS0FBSztBQUFsQixJQUFJQyxNQUFNLEdBQUdELFNBQUs7QUFDbEJFLE1BQU0sRUFBRSxLQUFLLElBQUlmLENBQUMsR0FBRyxDQUFDLElBQWU7RUFBUixJQUFBZ0IsTUFBQyxHQUFELENBQUM7RUFBTixJQUFBQyxNQUFNLEdBQU5qQixDQUFDLElBQUlnQixNQUFDO0VBQU4sSUFBQUMsTUFBTSxFQUF0QjtJQUFnQjtFQVN4QjtFQVRxQztJQUNuQyxLQUFLLElBQUlkLENBQUMsR0FBRyxDQUFDLElBQWU7TUFBUixJQUFBZSxNQUFDLEdBQUQsQ0FBQztNQUFOLElBQUFDLE1BQU0sR0FBTmhCLENBQUMsSUFBSWUsTUFBQztNQUFOLElBQUFDLE1BQU0sRUFBdEI7UUFBZ0I7TUFPaEI7TUFQNkI7UUFDakIsSUFBQUMsU0FBQyxHQUFELENBQUM7UUFBUCxJQUFBQyxTQUFPLEdBQVBsQixDQUFDLEtBQUtpQixTQUFDO1FBQVgsSUFBSUMsU0FBTyxFQUFFO1VBQ1g7VUFDQU4sTUFBTTtRQUNSLENBQUMsTUFBTTtVQUNJLElBQUFPLFVBQUksR0FBSixJQUFJO1VBQWJSLE1BQU0sR0FBR1E7VUFBVCxJQUFBQyxVQUFhLEdBQWJULE1BQWE7VUFBYlMsVUFBYTtRQUNmO01BQ0Y7TUFQd0IsSUFBQUMsTUFBRyxHQUFIckIsQ0FBQyxFQUFFO0lBTzNCO0VBQ0Y7RUFUZ0MsSUFBQXNCLE1BQUcsR0FBSHpCLENBQUMsRUFBRTtBQVNuQztBQUVlLElBQUEwQixTQUFJLEdBQUosSUFBSTtBQUFmLElBQUFDLFNBQWUsR0FBZmIsTUFBTSxLQUFLWSxTQUFJO0FBQW5CLElBQUlDLFNBQWUsRUFBRTtFQUNJLElBQUFDLFNBQWdFLEdBQWhFLGdFQUFnRTtFQUFqRixJQUFBQyxTQUFrRixHQUFsRixJQUFJbkIsWUFBWSxDQUFDa0IsU0FBZ0UsQ0FBQztFQUF4RixNQUFNQyxTQUFrRjtBQUMxRiIsImlnbm9yZUxpc3QiOltdfQ==) is doing something it shouldn't.

**Reply by nicolo-ribaudo** [reply]
Line 42 is missing a `!`
in the if condition :)
